DFC.Composite.Shell/Extensions/ServiceCollectionExtensions.cs

```diff
@@ -27,64 +39,253 @@ public static class ServiceCollectionExtensions
+                where TClientOptions : HttpClientOptions, new() =>
+            services
+                .Configure<TClientOptions>(configuration?.GetSection(configurationSectionName))
+                .AddHttpClient<TClient, TImplementation>()
+                .AddClientBuilder<TClientOptions>(retryPolicyKey, circuitBreakerPolicyKey);
+
+        public static IHttpClientBuilder AddClientBuilder<TClientOptions>(
+            this IHttpClientBuilder clientBuilder,
+            string retryPolicyKey,
+            string circuitBreakerPolicyKey)
+                where TClientOptions : HttpClientOptions, new() =>
+            clientBuilder
+                .ConfigureHttpClient((sp, options) =>
+                {
+                    var httpClientOptions = sp
+                        .GetRequiredService<IOptions<TClientOptions>>()
+                        .Value;
+                    options.BaseAddress = httpClientOptions.BaseAddress;
+                    options.Timeout = httpClientOptions.Timeout;
+                    options.DefaultRequestHeaders.Add(HeaderNames.Accept, MediaTypeNames.Text.Html);
+
+                    if (!string.IsNullOrWhiteSpace(httpClientOptions.ApiKey))
+                    {
+                        options.DefaultRequestHeaders.Add("Ocp-Apim-Subscription-Key", httpClientOptions.ApiKey);
+                    }
+                })
+                .ConfigurePrimaryHttpMessageHandler(() => new HttpClientHandler()
+                {
+                    // This prevents asp.net core from adding its own cookie values to the outgoing request
+                    UseCookies = false,
+                    AllowAutoRedirect = false,
+                })
+                .AddPolicyHandlerFromRegistry(retryPolicyKey)
+                .AddPolicyHandlerFromRegistry(circuitBreakerPolicyKey)
+                .AddHttpMessageHandler<UserAgentDelegatingHandler>()
+                .AddHttpMessageHandler<OriginalHostDelegatingHandler>()
+                .AddHttpMessageHandler<CompositeRequestDelegatingHandler>();
+
+        public static void ConfigureHttpClients(this IServiceCollection services, IConfiguration configuration)
+        {
+            var policyRegistry = services.AddPolicyRegistry();
+            var policyOptions = configuration?.GetSection(Constants.Policies)?.Get<PolicyOptions>();
+
+            if (policyOptions == null)
+            {
+                throw new Exception("Policy options could not be found");
+            }
+
+            AddHttpClientWithPolicies<INeo4JService, Neo4JService, VisitClientOptions>(
+                services,
+                policyRegistry,
+                $"{nameof(VisitClientOptions)}_{nameof(PolicyOptions.HttpRetry)}",
+                $"{nameof(VisitClientOptions)}_{nameof(PolicyOptions.HttpCircuitBreaker)}",
+                nameof(VisitClientOptions),
+                policyOptions,
+                configuration);
+
+            if (!services.AppRegistryRequestRegistered())
```

Comment:
Do we need this?

Reply:
It was a bit of a hack - I was using it for an integration test in my other big PR but can remove here